### ISSUE TITLE
Type inference based on return type

### ISSUE THREAD

**Issue by ImogenBits** [microsoft/pyright#8030]
It seems that type inference only uses types of arguments and ignores types provided by the context of the return type. For example:
```py
def decorator[T](value: T) -> Callable[[Callable[[T], T]], Callable[[T], T]]: ...

@decorator(1)
def something(a: float) -> float: ...
```
This raises an error because T is inferred to be int, not float. If T does not occur in the decorator's arguments, everything works as I'd expect since T is specialised only during the application to `something`.

Is this behaviour intentional? If so, is there some way to write a decorator like that without having to make it a generic class and explicitly specialising the types correctly (or typing the arguments as Any)?

**Comment by erictraut** [comment]
There is no type inference involved in your example. Pyright is evaluating the call expression `decorator(1)`, and the constraint solver is solving the type `T` based on the supplied argument. In this case, the argument is type `int`, so `T` takes on that value, and the return type is therefore `Callable[[Callable[[int], int]], Callable[[int], int]]`. 

Not surprisingly, mypy generates the same error in this case.

Pyright is working correctly here, so I don't consider this a bug. It's also not a valid enhancement request because changing the behavior as you are suggesting would make pyright non-conformant with the typing spec.

**Comment by ImogenBits** [comment]
Could you point me to the part of the typing spec that would be violated here? I've been trying to learn more about how and when exactly type vars are supposed to be specified and how e.g. the above example doesn't work, but it does work if the function is moved into the arguments, but I can't find anything about that in the typing spec.
